core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -833,33 +1106,44 @@ private[spark] class MapOutputTrackerWorker(conf: SparkConf) extends MapOutputTr
    *
    * (It would be nice to remove this restriction in the future.)
    */
-  private def getStatuses(shuffleId: Int, conf: SparkConf): Array[MapStatus] = {
-    val statuses = mapStatuses.get(shuffleId).orNull
-    if (statuses == null) {
-      logInfo("Don't have map outputs for shuffle " + shuffleId + ", fetching them")
+  private def getStatuses(
+      shuffleId: Int, conf: SparkConf): (Array[MapStatus], Array[MergeStatus]) = {
+    val mapOutputStatuses = mapStatuses.get(shuffleId).orNull
+    val mergeResultStatuses = mergeStatuses.get(shuffleId).orNull
+    if (mapOutputStatuses == null || (fetchMergeResult && mergeResultStatuses == null)) {
+      logInfo("Don't have map/merge outputs for shuffle " + shuffleId + ", fetching them")
       val startTimeNs = System.nanoTime()
       fetchingLock.withLock(shuffleId) {
-        var fetchedStatuses = mapStatuses.get(shuffleId).orNull
-        if (fetchedStatuses == null) {
-          logInfo("Doing the fetch; tracker endpoint = " + trackerEndpoint)
+        var fetchedMapStatuses = mapStatuses.get(shuffleId).orNull
+        if (fetchedMapStatuses == null) {
+          logInfo("Doing the map fetch; tracker endpoint = " + trackerEndpoint)
           val fetchedBytes = askTracker[Array[Byte]](GetMapOutputStatuses(shuffleId))
-          fetchedStatuses = MapOutputTracker.deserializeMapStatuses(fetchedBytes, conf)
-          logInfo("Got the output locations")
-          mapStatuses.put(shuffleId, fetchedStatuses)
+          fetchedMapStatuses = MapOutputTracker.deserializeOutputStatuses(fetchedBytes, conf)
+          logInfo("Got the map output locations")
+          mapStatuses.put(shuffleId, fetchedMapStatuses)
         }
-        logDebug(s"Fetching map output statuses for shuffle $shuffleId took " +
+        var fetchedMergeStatues = mergeStatuses.get(shuffleId).orNull
+        if (fetchMergeResult && fetchedMergeStatues == null) {
+          logInfo("Doing the merge fetch; tracker endpoint = " + trackerEndpoint)
+          val fetchedBytes = askTracker[Array[Byte]](GetMergeResultStatuses(shuffleId))
+          fetchedMergeStatues = MapOutputTracker.deserializeOutputStatuses(fetchedBytes, conf)
+          logInfo("Got the merge output locations")
+          mergeStatuses.put(shuffleId, fetchedMergeStatues)
+        }
+        logDebug(s"Fetching map/merge output statuses for shuffle $shuffleId took " +
           s"${TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTimeNs)} ms")
-        fetchedStatuses
+        (fetchedMapStatuses, fetchedMergeStatues)
       }
```

Comment:
That's not always true.
We currently fetch map status and merge status using 2 separate RPCs.
Although the fetching of these statuses is guarded by the lock, the initial check at line 1113 for these statuses being not null is out of the lock.
So, it would be possible that a task might see the map status being non-null while merge status being null.

We always need to fetch both map status and merge status together, either during the initial fetch or during fallback.
Combine both RPCs into 1 would increase the code complexity.
For now, the RPC just returns the pre-serialized bytes for either MapStatus array or MergeStatus array.
If we want to combine both into a single RPC, we would need to define additional RPC messages so that we can encode the 2 byte arrays for serialized MapStatus array and MergeStatus array together.
Combining both together does not seem to bring enough benefits.
We haven't observed any issue indicating Spark driver performance regression with doubling the number of RPCs for fetching shuffle statuses.
This would also help to keep code simpler.

Reply:
I would expect multiple rpc's to not be the preferred option given the impact on driver, but code simplicity needs to be balanced against that.
+CC @JoshRosen, @Ngone51 who last made changes here. Any thoughts on this ?

Reply:
One option could be replace `GetMergeResultStatuses` with `GetMapOutputAndMergeResultStatuses`.
That keeps non push based shuffle codepaths unchanged, and when push based shuffle is enabled, a single rpc handles the response : the code change would mirror what has been done for `GetMergeResultStatuses` already.

Thoughts @Ngone51, @Victsm, @venkata91 ?

Reply:
+1. This actually what I mean above.

Reply:
With the current RPC (`RpcCallContext`) mechanism with `MapOutputTracker`, we can only send one response as oppose to other RPC mechanisms with in Spark. If we have to combine getting both `MapStatuses` and `MergeStatuses` when push based shuffle is enabled, then we have couple of options:

1. Encode both MapStatuses and MergeStatuses in the same `Array[Byte]` getting returned from `serializedOutputStatus` with some encoding scheme like `length` in bytes of mapStatuses as first part and then mapStatuses similarly for mergeStatuses and in the `deserializeOutputStatuses` we have to decode it accordingly for the output of `GetMapOutputAndMergeResultStatuses` RPC call. This is some what not a cleaner approach as the client keeps the semantics of encoding/decoding of the byte array instead of the RPC layer itself. Although this is already being done wrt whether the mapStatuses are a `DIRECT` fetch or `BROADCAST` fetch. 
2. If not, we might need to make changes to `RpcCallContext` in order to respond with 2 byte arrays. This seems to be lot of additional overhead just for this purpose.

Any other suggestions? cc @Victsm @mridulm @Ngone51

Reply:
Gentle ping @Victsm @mridulm @Ngone51

Reply:
That is an implementation detail of what is response of `GetMapOutputAndMergeResultStatuses` right ?
It can simply be Array[Array[Byte]] (for example) - where result(0) is for `MapStatus` and result(1) is for `MergeStatus` - keeping everything else same ?

Reply:
Makes sense @mridulm Instead of `Array[Array[Byte]]` I used a tuple of `(Array[Byte], Array[Byte])`. cc @Victsm

Reply:
I am fine with Tuple as well.
+CC @Ngone51 in case you have any other thoughts.